napari/utils/progress.py

```diff
@@ -150,3 +151,68 @@ def progrange(*args, **kwargs):
 
     """
     return progress(range(*args), **kwargs)
+
+
+class cancelable_progress(progress):
+    """This class inherits from progress, providing the additional
+    ability to cancel expensive executions. When progress is
+    canceled by the user in the napari UI, two things will happen:
+
+    Firstly, the is_canceled attribute will become True, and the
+    for loop will terminate after the current iteration, regardless
+    of whether or not the iterator had more items.
+
+    Secondly, cancel_callback will be called, allowing the computation
+    to close resources, repair state, etc.
+
+    See napari.utils.progress and tqdm.tqdm API for valid args and kwargs:
+    https://tqdm.github.io/docs/tqdm/
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     def on_cancel():
+    ...         print("Canceled operation")
+    ...     for i in cancelable_progress(range(steps), cancel_callback=on_cancel):
+    ...         sleep(delay)
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        nest_under: Optional['progress'] = None,
+        cancel_callback: Optional[Callable] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        self.cancel_callback = cancel_callback
+        self.is_canceled = False
+
+        super().__init__(iterable, desc, total, nest_under, *args, **kwargs)
+
+    def __iter__(self) -> Iterator:
+        itr = super().__iter__()
+
+        def is_canceled(_):
+            if self.is_canceled:
+                # If we've canceled, run the callback and then notify takewhile
+                if self.cancel_callback:
+                    self.cancel_callback()
+                    # Perform additional cleanup for generators
+                    if isinstance(itr, Generator):
+                        itr.close()
```

Comment:
```suggestion
                    if isinstance(self.iterable, Generator):
                        self.iterable.close()
```

The `tqdm` exposes the inner iterable object on attribute `iterable`.